### PR TITLE
SQLite Update to get it closer to Postgres with ecto. 

### PR DESCRIPTION
**Pull request by jeregrine**
I'm bringing the discussion here, changes so far and work todo

- [x] DISTINCT is supported, just not column wise, need to check that distinct is called like MYSQL with `distinct: true` only
  - [x] remove order by code I added 
- [x] Converted JSON types to TEXT
- [x] Added Proper casing to REALS
- [x] Adding parens around sub commands for consistency with Postgres. (NOT union/except/intersect)
- [x] STARTED adding actual errors when users try to use some unsupported features. This is incomplete.
  - [x] DROP Index
  - [x] Unsupported column type
  - [x] Joins not supported with CTE
  - [x] LOCKs not supported in general, CTE's as well
  - [x] UPDATE with joins
  - [x] ALTER table stuff has lots of unsupported behaviors
- [x] Finish the "in" support tests
- [x] Check I didn't miss any tests from the OLD connection test.
- [x] Check the MYSQL tests to see if I have any others there that are relevant.

### CODE REVIEW

**Comment by warmwaffles** [comment]
Exciting.

**Comment by warmwaffles** [comment]
@jeregrine how's this coming along?

File: lib/ecto/adapters/sqlite3/connection.ex [hunk 1]
```diff
@@ -1359,8 +1397,13 @@ defmodule Ecto.Adapters.SQLite3.Connection do
   def expr({fun, _, args}, sources, query) when is_atom(fun) and is_list(args) do
     {modifier, args} =
       case args do
-        [rest, :distinct] -> {"DISTINCT ", [rest]}
-        _ -> {[], args}
+        [_rest, :distinct] ->
+          raise Ecto.QueryError,
+            query: query,
+            message: "Distinct not supported in expressions"
```
Comment (ruslandoga) [inline lib/ecto/adapters/sqlite3/connection.ex:1403]:
```sql
sqlite> create table example(a integer);
sqlite> insert into example(a) values(1), (1), (2);
sqlite> select count(distinct a) from example;
2
```

File: lib/ecto/adapters/sqlite3/connection.ex [hunk 2]
```diff
@@ -985,6 +1012,12 @@ defmodule Ecto.Adapters.SQLite3.Connection do
         source: source,
         hints: hints
       } ->
+        if hints != [] do
+          raise Ecto.QueryError,
+            query: query,
+            message: "join hints are not supported by SQLite3"
+        end
+
```
Comment (ruslandoga) [inline lib/ecto/adapters/sqlite3/connection.ex:1020]: Should `Enum.map(hints, &[?\s | &1]),` be removed from the iolist below?

File: lib/ecto/adapters/sqlite3/connection.ex (hunk 1 above)
Comment (jeregrine) [reply lib/ecto/adapters/sqlite3/connection.ex:1381]: It sort of support's them not in the same way that postgres and ecto support them. I started work adding them into this following the mysql ecto adapter approach but threw it out cause it got hairy. Probably worth a second look in a further PR though.

File: lib/ecto/adapters/sqlite3/connection.ex (hunk 2 above)
Comment (jeregrine) [reply lib/ecto/adapters/sqlite3/connection.ex:998]: good catch!

**Comment by jeregrine** [comment]
@warmwaffles I fixed the merge conflicts but integration tests will continue to fail until a new ecto version is cut with this in it https://github.com/elixir-ecto/ecto/pull/4083

**Comment by warmwaffles** [comment]
Ah no problem then. Let's wait for them to cut a new version. There shouldn't be too many merge conflicts until then.

**Comment by warmwaffles** [comment]
I want to give you all the credit for the work you did here. When I squash merge #107, it will consolidate under @fredwu's name.

What I can do is just bump to ecto 3.10, then you rebase this on top. Then @fredwu rebase his work after we merge your work @jeregrine.

How does that sound?

**Comment by warmwaffles** [comment]
Actually @jeregrine let me just squash merge this and @fredwu can just rebase on top.

**Pull request merged**